[Metabot] Sessions

sloansparger opened the pull request:
### Description

Implements sessions on the FE per the spec: https://www.notion.so/metabase/Get-v1-LLM-Metabot-complete-enough-for-stats-and-more-tool-building-10d69354c90180c4aa35ee5a83a39c40?pvs=4#12e69354c90180ec816cf9367e8e1b85
- creates a new session id when opened, clears it out when closed
- metabot will auto-close after 30 seconds


### Checklist

- [x] Tests have been added/updated to cover changes in this PR

sloansparger committed:
base e2e tests

sloansparger committed:
unit tests

sloansparger committed:
fe implementation for sessions

johnswanson committed:
BE: Send back a `session_id` if missing

sloansparger committed:
change to uuid

johnswanson committed:
send the session id along to ai-service

iethree commented on enterprise/frontend/src/metabase-enterprise/metabot/state/actions.ts:
```diff
@@ -69,10 +72,25 @@ export const sendMessageRequest = createAsyncThunk(
+    let sessionId = getMetabotSessionId(getState() as any);
+
+    // should not be needed, but just in case the value got unset
+    if (!sessionId) {
+      console.warn(
+        "Metabot has no session id while open, this should never happen",
+      );
```
I don't love this, but I guess it can't hurt 🤔

iethree commented on enterprise/frontend/src/metabase-enterprise/metabot/state/reducer.ts:
```diff
@@ -9,20 +10,22 @@ import type {
 
 export interface MetabotState {
-  isProcessing: boolean;
-  userMessages: string[];
   confirmationOptions: Record<string, MetabotReaction[]> | undefined;
+  isProcessing: boolean;
   lastSentContext: MetabotChatContext | undefined;
   lastHistoryValue: MetabotHistory | undefined;
+  sessionId: string | undefined;
+  userMessages: string[];
   visible: boolean;
 }
```
maybe this can find a better home?

sloansparger replied:
Where the Metabot session id lives or the state type definition?

iethree replied:
the state type

sloansparger replied:
I have it on my task list to rework the enterprise state, so I'll look at moving it as part of that change.

sloansparger committed:
pr feedback

johnswanson committed:
require `session_id` to be passed, and be a UUID

dpsutton committed:
backend: send a dummy session id to unbreak stats (#49894)

* backend: send a dummy session id to unbreak stats

* preserve session-id across calls

one shortcoming: it does not preserve the session id across FE
invocations. Need to send it back and the FE parrot it back, and the
envelope create function should reuse that when present

```
{
  "messages" : [ {
    "role" : "user",
    "content" : "filter to where the state is CA"
  } ],
  "context" : {...}
  "tools" : [,,,],
  "instance_info" : {,,,},
  "user_id" : 1,
  "session_id" : "6e67dafe-30e6-4b0a-8387-5dc5968e7f28" // <----
}

"LLM -----------------> BE"
{
  "message" : {
    "content" : "",
    "role" : "assistant",
    "tool_calls" : [ {
      "id" : "call_NpA5WEA2gzrRDWt7pC8i0GZR",
      "name" : "change_query",
      "arguments" : "{\"relative_date_filters\":null,\"specific_date_filters\":null,\"boolean_filters\":null,\"breakouts\":null,\"string_filters\":[{\"column\":\"State\",\"operator\":\"=\",\"value\":\"CA\"}],\"order_bys\":null,\"limits\":null,\"number_filters\":null,\"aggregations\":null}"
    } ]
  },
  "metadata" : {
    "usage" : {
      "model" : "gpt-4o-mini-2024-07-18",
      "total" : 7404,
      "prompt" : 7335,
      "completion" : 69,
      "estimated_costs" : 0.0011519999999999998
    }
  }
}

"BE -----------------> LLM"
{
  "messages" : [ {
    "role" : "user",
    "content" : "filter to where the state is CA"
  }, {
    "content" : "",
    "role" : "assistant",
    "tool_calls" : [ {
      "id" : "call_NpA5WEA2gzrRDWt7pC8i0GZR",
      "name" : "change_query",
      "arguments" : "{\"order-bys\":null,\"specific-date-filters\":null,\"breakouts\":null,\"boolean-filters\":null,\"relative-date-filters\":null,\"string-filters\":[{\"column\":\"State\",\"operator\":\"=\",\"value\":\"CA\"}],\"number-filters\":null,\"limits\":null,\"aggregations\":null}"
    } ]
  }, {
    "role" : "tool",
    "tool_call_id" : "call_NpA5WEA2gzrRDWt7pC8i0GZR",
    "content" : "success"
  } ],
  "context" : {},
  "tools" : []..,
  "instance_info" : {,,,}
  },
  "user_id" : 1,
  "session_id" : "6e67dafe-30e6-4b0a-8387-5dc5968e7f28" // <---
}
```

This session_id above is missing in the BE -> FE and then on subsequent FE -> BE payloads

---------

Co-authored-by: dan sutton <dan@dpsutton.com>

sloansparger committed:
Merge branch 'metabot-v3-second-try' into metabot-v3-sessions

sloansparger committed:
right some wrongs from merge conflict resolutions

sloansparger committed:
fix

sloansparger committed:
fix
